### DISCUSSION TITLE
Is pgvecto.rs 0.3 already supported?

### DISCUSSION THREAD

**Discussion by fdw** [immich-app/immich#11091] [Q&A]
The Docker compose files still list v0.2, but I have no idea if that is on purpose or not.

**Comment by mertalev** [comment]
Immich only accepts 0.2.x at the moment. 0.2.1 is safe to upgrade to, but 0.3.0 requires some testing. The pgvector compatibility mode in particular seemed to not work for a user until they downgraded.

**Reply by Aetf** [reply]
I accidentally upgraded mine from 0.2.0 to 0.3.0 :facepalm:. Really shouldn't be doing version upgrades late night...

Now immich refuses to connect. And apparently the extension can't be easily downgraded. I restarted the db with 0.2.0 and tried to `ALTER EXTENSION vectors UPDATE` and got 
```
extension "vectors" has no update path from version "0.3.0" to version "0.2.0"
```

Any way to recover? Otherwise I guess I would have to have immich down before it supports 0.3...

**Reply by mertalev** [reply]
In the case of the person I helped, the downgrade path was to drop `face_index` and `clip_index`, change the embedding columns to `real array` and drop the extension. Then once they downgraded their image to 0.2.*, create the extension, change the column types back and recreate the indices.

I can also take another look at 0.3.0 and see if we can support it in the next release.

**Reply by Aetf** [reply]
Thanks for the prompt reply! I'll wait for 0.3.0 support then. My instance has been broken for other reasons for a while now anyway...

**Reply by Aetf** [reply]
I finally did the downgrade. With some trial and error, my instance is up and running!

Documenting the commands here:
After connecting using psql
```
# database image is already downgraded to 0.2

DROP INDEX face_index;
DROP INDEX clip_index;
ALTER TABLE asset_faces ALTER COLUMN embedding TYPE real array;
ALTER TABLE smart_search ALTER COLUMN embedding TYPE real array;
DROP EXTENSION vectors;
CREATE EXTENSION vectors;
ALTER EXTENSION vectors UPDATE;
SELECT pgvectors_upgrade();

# restart the database instance

ALTER TABLE smart_search ALTER COLUMN embedding TYPE vector(512);
ALTER TABLE asset_faces ALTER COLUMN embedding TYPE vector(512);
SET vectors.pgvector_compatibility=on;
CREATE INDEX face_index ON asset_faces USING hnsw (embedding vector_cosine_ops) WITH (ef_construction = 300, m = 16);
CREATE INDEX clip_index ON smart_search USING hnsw (embedding vector_cosine_ops) WITH (ef_construction = 300, m = 16);
GRANT SELECT ON TABLE pg_vector_index_stat TO PUBLIC;
GRANT USAGE ON SCHEMA vectors TO your-db-user-name;
```

**Reply by mertalev** [reply]
Awesome, glad to hear it worked! And thanks for sharing the steps you took!

Incidentally, we'll probably support 0.3.0 in the next release. I didn't have any issues with it during my testing with #11559.

**Comment by mmomjian** [comment]
Supported
